persistent-postgresql/Database/Persist/Postgresql.hs

```diff
@@ -1151,6 +1153,38 @@ instance PersistConfig PostgresConf where
         addPass     = maybeAddParam "password" "PGPASS"
         addDatabase = maybeAddParam "dbname"   "PGDATABASE"
 
+refName :: DBName -> DBName -> DBName
+refName (DBName table) (DBName column) =
+    let overhead = T.length $ T.concat ["_", "_fkey"]
+        (fromTable, fromColumn) = shortenNames overhead (T.length table, T.length column)
+    in DBName $ T.concat [T.take fromTable table, "_", T.take fromColumn column, "_fkey"]
+
+    where
+
+      -- Postgres automatically truncates too long foreign keys to a combination of
+      -- truncatedTableName + "_" + truncatedColumnName + "_fkey"
+      -- This works fine for normal use cases, but it creates an issue for Persistent
+      -- Because after running the migrations, Persistent sees the truncated foreign key constraint
+      -- doesn't have the expected name, and suggests that you migrate again
+      -- To workaround this, we copy the Postgres truncation approach before sending foreign key constraints to it.
+      --
+      -- I believe this will also be an issue for extremely long table names, 
+      -- but it's just much more likely to exist with foreign key constraints because they're usually tablename * 2 in length
+
+      -- Approximation of the algorithm Postgres uses to truncate identifiers
+      -- See makeObjectName https://github.com/postgres/postgres/blob/5406513e997f5ee9de79d4076ae91c04af0c52f6/src/backend/commands/indexcmds.c#L2074-L2080
+      shortenNames :: Int -> (Int, Int) -> (Int, Int)
+      shortenNames overhead (x, y) =
+        if | x + y + overhead <= maximumIdentifierLength -> (x, y)
+           | x > y -> shortenNames overhead (x - 1, y)
+           | otherwise -> shortenNames overhead (x, y - 1)
```

Comment:
This can be accomplished with ordinary guards.

```haskell
shortenNames overhead (x, y)
    | x + y + overhead <= maximumIdentifierLenght = (x, y)
    | x > y = shortenNames overhead (x - 1, y)
    | otherwise = shortenNames overhead (x, y - 1)
```